cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -69,10 +69,11 @@ impl ArgAssigner for Args {
     fn assign(&mut self, arg: &AbiParam) -> ArgAction {
         let ty = arg.value_type;
 
-        // Check for a legal type.
-        // We don't support SIMD yet, so break all vectors down.
+        // Vectors should stay in vector registers TODO perhaps this should be conditional on some ISA feature (e.g. SSE) and if not present do a ValueConversion
         if ty.is_vector() {
-            return ValueConversion::VectorSplit.into();
+            let reg = FPR.unit(self.fpr_used);
+            self.fpr_used += 1;
+            return ArgumentLoc::Reg(reg).into();
```

Comment:
@bnjbvr, as we discussed, I tried to use some of the FPR code from below and discovered some bizarre behavior when I run the `legalize-simd.clif` test:

```
> function %test_simd(i64 fp [%rbp]) -> i32x4 [%xmm0], i64 fp [%rbp] fast {
Missed #0: \bv2 = bitcast\.i32x4 v0\b
>     ss0 = incoming_arg 16, offset -16
> 
>                                 ebb0(v3: i64 [%rbp]):
> [RexOp1pushq#50]                    x86_push v3
> [RexOp1copysp#8089]                 copy_special %rsp -> %rbp
> [RexOp1pu_id#b8,%rax]               v0 = iconst.i32 42
> [RexMp2frurm#56e,%xmm0]             v2 = bitcast.i32x4 v0
> [Mp2r_ib_unsigned#570,%xmm0]        v1 = x86_pshuf v2, 0
> [RexOp1popq#58,%rbp]                v4 = x86_pop.i64 
> [Op1ret#c3]                         return v1, v4
> }
```

Happily the `v1` return value seems to be assigned to `%XMM0` but I don't understand how `%rbp` has now become both an input parameter (the EBB as written has no parameters) and a return value.

Reply:
I think `fast` defines a calling convention which generates its prologue (push RBP; move RSP to RBP) and epilogue (pop RBP). So it's probably expected.

Reply:
I moved to the `baldrdash` calling convention and altered the test a bit and now all the tests pass!

Reply:
Great job!